### DISCUSSION TITLE
posts/hello-world/

### DISCUSSION THREAD

**Discussion by giscus** [BYJRK/byjrk.github.io#1] [Announcements]
# posts/hello-world/

欢迎来到我的博客！

https://blog.coldwind.top/posts/hello-world/

**Comment by BYJRK** [comment]
沙发自己坐

**Reply by Lucifer201** [reply]
a

**Comment by MUSIC-DIE** [comment]
加油！！！支持！！！三连！！！

**Reply by xiaososa** [reply]
支持大佬，边给大佬“充电”边给自己充电~~！！！！！

**Comment by xiaososa** [comment]
坐一个沙发~

**Comment by SlimeNull** [comment]
好耶! 支持大佬!

**Comment by GitClound** [comment]
我想了解一下这个博客怎么搭建起来的？希望可以分享分享

**Reply by BYJRK** [reply]
博客搭建所使用的框架以及主题都在最下方可以看到。Hugo的入门难度可能相对高一些，你可以看看官方文档的[入门教程](https://gohugo.io/getting-started/quick-start/)

**Reply by BYJRK** [reply]
有一位小伙伴参考我的博客搭建了他自己的，并且写了个教程。转需：https://lxymahatma.github.io/p/how-to-build-blog/

**Comment by mxj12315** [comment]
大佬 膜拜一下

**Comment by XVCoder** [comment]
大佬不自己撸一个博客管理系统吗 :dog:

**Reply by BYJRK** [reply]
精力有限

**Comment by dawnlyn** [comment]
我最近在学习Avalonia，想要做一个管理B站收藏夹(不止B站)的小程序，但是我现在遇到了一个问题：我的布局中存在一个Header，是继承UserControl的，然后我现在需要点击其中的按钮并打开一个弹窗，但是我不太清楚如何操作。我跟着官方教程完成了那个MusicStore，最近正在考虑是否需要从ReactiveUI转到CommunityToolkit.MVVM

**Reply by BYJRK** [reply]
我个人肯定是更推荐Toolkit.Mvvm的，因为它更加简单，符合一般MVVM的做法，上手难度比ReactiveUI低。一般情况下，你可能根本用不上ReactiveUI的功能，或者发挥出它的优势的，除非你确信你需要响应式编程

**Reply by dawnlyn** [reply]
其实，我昨天试了一下，用Toolkit.Mvvm来重写官方的哪个MusicStore，然后我就发现弹窗那一块写着好难受，Messager在复杂情况下完全搞不懂....最后还是把两个结合了....

**Reply by SlimeNull** [reply]
我感觉用 Custom Control 用的更多, UserControl 我只有封装页面的时候用.

**Reply by BYJRK** [reply]
@SlimeNull UserControl除了用于封装整个界面（一般用于导航），也常用于封装一个小模块，比如一个包含了Label、TextBox、Button的可复用模块。

**Reply by SlimeNull** [reply]
> @SlimeNull UserControl除了用于封装整个界面（一般用于导航），也常用于封装一个小模块，比如一个包含了Label、TextBox、Button的可复用模块。

我更喜欢使用 CustomControl去封装，反正就是自定义模板的这种

**Reply by BYJRK** [reply]
虽然都能实现效果，但这两个的用法和适用情形上还是存在一些区别的。一般来说，如果想要组合现有控件，且不和样式打交道，或者说上面的控件都用默认样式，还可以有一些依赖属性，那么用 UserControl 是非常合适的。CustomControl 不像 UserControl 那样直接有一个 XAML 文件，而是需要写 Template，有时候并不是那么必要。

**Reply by SlimeNull** [reply]
复杂的控件, 说白了就是 CustomControl 来的, 由基本的元素组合而来. 所以我更倾向于使用 CustomControl.

**Reply by BYJRK** [reply]
你还是这么固执😂

**Reply by SlimeNull** [reply]
因为确实, 我封装的乱七八糟组件, 就是用 CustomControl, 而且我也认为 CustomControl 好用. 其实主要是受到 WPF 影响. 在 WPF 中, UserControl 你搞 Binding, 就得想办法设置一个 DataContext, 或者给根控件命名, 但是 CustomControl 可以用 TemplateBinding. 而且 UserControl 写完之后, 假如你要弄成库, 别人也不能改模板, 拓展性不太行.

**Reply by XVCoder** [reply]
@SlimeNull
用法不一样吧，看看这个→ [WPF Tutorial | UserControls vs. CustomControls](https://www.wpftutorial.net/CustomVsUserControl.html)

**Reply by SlimeNull** [reply]
> @SlimeNull 用法不一样吧，看看这个→ [WPF Tutorial | UserControls vs. CustomControls](https://www.wpftutorial.net/CustomVsUserControl.html)

这教程上说的不好. CustomControl 组合控件也非常方便. 就是写模板, 还能 TemplateBinding. 你可以看 GroupBox, 它不就是组合 header 和 body, 然后 header 和 body 都支持自定义的? 你可以参考我写的 OpenGptChat 项目中的 ChatBubble 控件, 也就是聊天气泡. 包含一个标头, 边框, 以及边框中的内容. 并且支持 Markdown 渲染.

向外暴露 Title 依赖属性以及 MessageText 依赖属性, 我就可以在模板中使用 TemplateBinding 将它绑定到 TextBlock 上, 然后还可以通过 TemplateBinding Visibility 之类的操作, 实现向外暴露 "EnableMarkdownRendering" 属性, 然后外部对其赋值.

最主要的还是, UserControl 不能用 TemplateBinding, 只能用 Binding, 而 WPF 的 Binding 并没有那么好用, 所以我更喜欢用 CustomControl

> C# 代码: [OpenGptChat | ChatBubble.cs](https://github.com/SlimeNull/OpenGptChat/blob/main/OpenGptChat/Controls/ChatBubble.cs)
> XAML [OpenGptChat | Generic.xaml](https://github.com/SlimeNull/OpenGptChat/blob/main/OpenGptChat/Themes/Generic.xaml)

**Reply by SlimeNull** [reply]
这不是我直接没根据的判断 CustomControl 好用, 而是在我两个方式都试过之后, 确实发现 CustomControl 要更好用的.

**Reply by BYJRK** [reply]
“User”想表达什么意思呢？显然就是把这个东西限制在自己项目中了。你搞了一个只能自己用的小组合控件也好，搞了一整个页面也好，都只能你自己用，不能或不适合分享给别人（别的项目）的。想分享肯定要用CustomControl。至于TemplateBinding，给UC添加DP不也可以绑吗？所以我们想讨论的是它们俩啥时候用谁更合适更简便，而不是在说CC不好用、难用、或者不如UC还是怎么的。

**Reply by SlimeNull** [reply]
UserControl 怎么绑定呢, 我目前想过两种方案, 第一种是在非根控件下设定 DataContext 为 this, 这样下面所有 Binding 都是直接绑定到当前 UserControl 的依赖属性上. 第二种是为根控件设定名称, 然后所有的绑定都使用 `ElementName=root,Path=XXX` 这样的写法.
还有别的方案吗?

**Reply by SlimeNull** [reply]
是因为我找不到 UserControl 中更优雅的绑定写法, 所以才觉得 CustomControl 好用. (其实也就只是因为绑定这一个原因)
